In [1]:
print("Hello, World!")

Hello, World!


### This is my capstone project for CFA level 1 Practical Skills Module.
Portfolio optimization is the process of selecting the best portfolio out of a set of available 
portfolios with the objective of maximizing returns and reducing risk.

Before we get started, we need to define the following terms:
* Open: Price at which a security first trades when stock market opens. 
* High: Highest price at which a given stock traded during a regular trading session. 
* Low: Lowest price at which a given stock traded during a regular trading session. 
* Close: Last price at which a given stock traded during a regular trading session. 
* Volume: Number of shares traded between the market open and close.
* Adj. Close: Adjusted stock close price after considering stock splits and dividends.

In [2]:
#import libraries
!pip install pandas
import pandas as pd
import numpy as ny

#datetime module comes pre installed in python
import datetime as dt

Defaulting to user installation because normal site-packages is not writeable


In [3]:
amazon_df = pd.read_csv('Amazon.csv')
jpm_df = pd.read_csv('JPM.csv')

In [4]:
print(amazon_df.head(10))
print(jpm_df.head())

        Date       Open       High        Low      Close     Volume  Adj Close
0   1/2/2018  58.599998  59.500000  58.525501  59.450500   53890000  59.450500
1   1/3/2018  59.415001  60.274502  59.415001  60.209999   62176000  60.209999
2   1/4/2018  60.250000  60.793499  60.233002  60.479500   60442000  60.479500
3   1/5/2018  60.875500  61.457001  60.500000  61.457001   70894000  61.457001
4   1/8/2018  61.799999  62.653999  61.601501  62.343498   85590000  62.343498
5   1/9/2018  62.845001  62.966499  62.088001  62.634998   73226000  62.634998
6  1/10/2018  62.257500  62.716499  61.861500  62.716499   53720000  62.716499
7  1/11/2018  62.987000  63.838501  62.823002  63.834000   62500000  63.834000
8  1/12/2018  63.669498  65.288002  63.669498  65.260002  108874000  65.260002
9  1/16/2018  66.150002  66.997002  64.614998  65.242996  144414000  65.242996
        Date       Open       High        Low      Close  Adj Close    Volume
0  7/14/2017  90.809998  92.610001  90.580002  92.250

In [5]:
#checking the number of null values in the data frame
amazon_df.isnull().sum()

Date         0
Open         0
High         0
Low          0
Close        0
Volume       0
Adj Close    0
dtype: int64

In [6]:
#checking the null values in JPM data frame
jpm_df.isnull().sum()

Date         0
Open         0
High         0
Low          0
Close        0
Adj Close    0
Volume       0
dtype: int64

No null values were found in both the data frames

In [7]:
#Amazon's info values
amazon_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1250 entries, 0 to 1249
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       1250 non-null   object 
 1   Open       1250 non-null   float64
 2   High       1250 non-null   float64
 3   Low        1250 non-null   float64
 4   Close      1250 non-null   float64
 5   Volume     1250 non-null   int64  
 6   Adj Close  1250 non-null   float64
dtypes: float64(5), int64(1), object(1)
memory usage: 68.5+ KB


In [8]:
amazon_df.describe()

Open         High          Low        Close        Volume  \
count  1250.000000  1250.000000  1250.000000  1250.000000  1.250000e+03   
mean    120.149168   121.589924   118.536881   120.067865  8.670803e+07   
std      35.656990    36.040213    35.226187    35.591723  4.077817e+07   
min      58.599998    59.500000    58.525501    59.450500  1.762600e+07   
25%      89.207373    89.902502    88.098499    89.128374  5.897078e+07   
50%     108.294998   111.704998   106.855003   108.734501  7.515300e+07   
75%     158.389877   160.050995   156.275372   158.111000  1.027046e+08   
max     187.199997   188.654007   184.839493   186.570496  3.113460e+08   

         Adj Close  
count  1250.000000  
mean    120.067865  
std      35.591723  
min      59.450500  
25%      89.128374  
50%     108.734501  
75%     158.111000  
max     186.570496

In [9]:
#Doing the same for JPM file
jpm_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1368 entries, 0 to 1367
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       1368 non-null   object 
 1   Open       1368 non-null   float64
 2   High       1368 non-null   float64
 3   Low        1368 non-null   float64
 4   Close      1368 non-null   float64
 5   Adj Close  1368 non-null   float64
 6   Volume     1368 non-null   int64  
dtypes: float64(5), int64(1), object(1)
memory usage: 74.9+ KB


In [10]:
jpm_df.describe()

Open         High          Low        Close    Adj Close  \
count  1368.000000  1368.000000  1368.000000  1368.000000  1368.000000   
mean    120.672434   121.882880   119.455768   120.650168   112.181658   
std      21.824789    21.913014    21.709343    21.808955    23.815861   
min      81.559998    83.750000    76.910004    79.029999    72.579208   
25%     104.732498   105.699997   103.730003   104.757502    93.908926   
50%     114.514999   115.405003   113.439999   114.535000   102.741505   
75%     134.887497   136.447495   133.935001   135.292496   128.463585   
max     172.710007   172.960007   170.539993   171.779999   166.523987   

             Volume  
count  1.368000e+03  
mean   1.441361e+07  
std    6.503746e+06  
min    3.220500e+06  
25%    1.022870e+07  
50%    1.285545e+07  
75%    1.635760e+07  
max    5.441880e+07

From the description we can observe Amazon has the maximum stock prices and JPM has better minimum stock prices.
Amazon ranges from 50s to 180s whereas JPM stock prices ranges from 80s to 170s. The mean is similar in both cases.

### Calculating Percentage Daily returns for a stock

In [11]:
#creating a daily returns column in the data frames
amazon_df['Daily Return'] = amazon_df['Adj Close'].pct_change(1) * 100
jpm_df['Daily Return'] = jpm_df['Adj Close'].pct_change(1) * 100

In [12]:
#displaying the data frames to see the new column
amazon_df

Date       Open       High        Low      Close     Volume  \
0       1/2/2018  58.599998  59.500000  58.525501  59.450500   53890000   
1       1/3/2018  59.415001  60.274502  59.415001  60.209999   62176000   
2       1/4/2018  60.250000  60.793499  60.233002  60.479500   60442000   
3       1/5/2018  60.875500  61.457001  60.500000  61.457001   70894000   
4       1/8/2018  61.799999  62.653999  61.601501  62.343498   85590000   
...          ...        ...        ...        ...        ...        ...   
1245  12/12/2022  89.209999  90.580002  87.870003  90.550003   61999800   
1246  12/13/2022  95.230003  96.250000  90.519997  92.489998  100212000   
1247  12/14/2022  92.500000  93.459999  89.870003  91.580002   70298000   
1248  12/15/2022  89.889999  89.970001  87.470001  88.449997   84802900   
1249  12/16/2022  88.269997  89.349998  86.730003  87.860001  146044000   

      Adj Close  Daily Return  
0     59.450500           NaN  
1     60.209999      1.277531  
2     60.479500      0.447601  
3     61.457001      1.616252  
4     62.343498      1.442468  
...         ...           ...  
1245  90.550003      1.638800  
1246  92.489998      2.142457  
1247  91.580002     -0.983886  
1248  88.449997     -3.417782  
1249  87.860001     -0.667039  

[1250 rows x 8 columns]

In [13]:
jpm_df

Date        Open        High         Low       Close   Adj Close  \
0      7/14/2017   90.809998   92.610001   90.580002   92.250000   79.338943   
1      7/17/2017   91.820000   91.989998   91.250000   91.389999   78.599312   
2      7/18/2017   90.449997   91.580002   90.320000   91.070000   78.324104   
3      7/19/2017   91.339996   91.620003   91.000000   91.199997   78.435898   
4      7/20/2017   91.150002   91.720001   90.900002   91.199997   78.435898   
...          ...         ...         ...         ...         ...         ...   
1363  12/12/2022  132.399994  134.649994  131.600006  134.210007  134.210007   
1364  12/13/2022  136.889999  137.089996  133.080002  134.080002  134.080002   
1365  12/14/2022  133.779999  135.710007  132.759995  133.410004  133.410004   
1366  12/15/2022  131.149994  132.080002  129.050003  130.100006  130.100006   
1367  12/16/2022  129.149994  129.899994  128.429993  129.289993  129.289993   

        Volume  Daily Return  
0     22235200           NaN  
1     14374200     -0.932243  
2     14719400     -0.350140  
3     11651200      0.142732  
4     11561700      0.000000  
...        ...           ...  
1363   8841600      1.551152  
1364  10025400     -0.096867  
1365   9966100     -0.499700  
1366  12087800     -2.481072  
1367  21126400     -0.622608  

[1368 rows x 8 columns]

In [14]:
amazon_df['Daily Return'].replace(ny.nan, 0, inplace = True)
jpm_df['Daily Return'].replace(ny.nan, 0, inplace = True)

C:\Users\willy\AppData\Local\Temp\ipykernel_6344\3837966503.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  amazon_df['Daily Return'].replace(ny.nan, 0, inplace = True)
C:\Users\willy\AppData\Local\Temp\ipykernel_6344\3837966503.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy

In [15]:
amazon_df

Date       Open       High        Low      Close     Volume  \
0       1/2/2018  58.599998  59.500000  58.525501  59.450500   53890000   
1       1/3/2018  59.415001  60.274502  59.415001  60.209999   62176000   
2       1/4/2018  60.250000  60.793499  60.233002  60.479500   60442000   
3       1/5/2018  60.875500  61.457001  60.500000  61.457001   70894000   
4       1/8/2018  61.799999  62.653999  61.601501  62.343498   85590000   
...          ...        ...        ...        ...        ...        ...   
1245  12/12/2022  89.209999  90.580002  87.870003  90.550003   61999800   
1246  12/13/2022  95.230003  96.250000  90.519997  92.489998  100212000   
1247  12/14/2022  92.500000  93.459999  89.870003  91.580002   70298000   
1248  12/15/2022  89.889999  89.970001  87.470001  88.449997   84802900   
1249  12/16/2022  88.269997  89.349998  86.730003  87.860001  146044000   

      Adj Close  Daily Return  
0     59.450500      0.000000  
1     60.209999      1.277531  
2     60.479500      0.447601  
3     61.457001      1.616252  
4     62.343498      1.442468  
...         ...           ...  
1245  90.550003      1.638800  
1246  92.489998      2.142457  
1247  91.580002     -0.983886  
1248  88.449997     -3.417782  
1249  87.860001     -0.667039  

[1250 rows x 8 columns]

In [16]:
jpm_df

Date        Open        High         Low       Close   Adj Close  \
0      7/14/2017   90.809998   92.610001   90.580002   92.250000   79.338943   
1      7/17/2017   91.820000   91.989998   91.250000   91.389999   78.599312   
2      7/18/2017   90.449997   91.580002   90.320000   91.070000   78.324104   
3      7/19/2017   91.339996   91.620003   91.000000   91.199997   78.435898   
4      7/20/2017   91.150002   91.720001   90.900002   91.199997   78.435898   
...          ...         ...         ...         ...         ...         ...   
1363  12/12/2022  132.399994  134.649994  131.600006  134.210007  134.210007   
1364  12/13/2022  136.889999  137.089996  133.080002  134.080002  134.080002   
1365  12/14/2022  133.779999  135.710007  132.759995  133.410004  133.410004   
1366  12/15/2022  131.149994  132.080002  129.050003  130.100006  130.100006   
1367  12/16/2022  129.149994  129.899994  128.429993  129.289993  129.289993   

        Volume  Daily Return  
0     22235200      0.000000  
1     14374200     -0.932243  
2     14719400     -0.350140  
3     11651200      0.142732  
4     11561700      0.000000  
...        ...           ...  
1363   8841600      1.551152  
1364  10025400     -0.096867  
1365   9966100     -0.499700  
1366  12087800     -2.481072  
1367  21126400     -0.622608  

[1368 rows x 8 columns]

In [17]:
amazon_df = amazon_df.round(2) 
amazon_df

Date   Open   High    Low  Close     Volume  Adj Close  \
0       1/2/2018  58.60  59.50  58.53  59.45   53890000      59.45   
1       1/3/2018  59.42  60.27  59.42  60.21   62176000      60.21   
2       1/4/2018  60.25  60.79  60.23  60.48   60442000      60.48   
3       1/5/2018  60.88  61.46  60.50  61.46   70894000      61.46   
4       1/8/2018  61.80  62.65  61.60  62.34   85590000      62.34   
...          ...    ...    ...    ...    ...        ...        ...   
1245  12/12/2022  89.21  90.58  87.87  90.55   61999800      90.55   
1246  12/13/2022  95.23  96.25  90.52  92.49  100212000      92.49   
1247  12/14/2022  92.50  93.46  89.87  91.58   70298000      91.58   
1248  12/15/2022  89.89  89.97  87.47  88.45   84802900      88.45   
1249  12/16/2022  88.27  89.35  86.73  87.86  146044000      87.86   

      Daily Return  
0             0.00  
1             1.28  
2             0.45  
3             1.62  
4             1.44  
...            ...  
1245          1.64  
1246          2.14  
1247         -0.98  
1248         -3.42  
1249         -0.67  

[1250 rows x 8 columns]

In [18]:
jpm_df = jpm_df.round(2)
jpm_df

Date    Open    High     Low   Close  Adj Close    Volume  \
0      7/14/2017   90.81   92.61   90.58   92.25      79.34  22235200   
1      7/17/2017   91.82   91.99   91.25   91.39      78.60  14374200   
2      7/18/2017   90.45   91.58   90.32   91.07      78.32  14719400   
3      7/19/2017   91.34   91.62   91.00   91.20      78.44  11651200   
4      7/20/2017   91.15   91.72   90.90   91.20      78.44  11561700   
...          ...     ...     ...     ...     ...        ...       ...   
1363  12/12/2022  132.40  134.65  131.60  134.21     134.21   8841600   
1364  12/13/2022  136.89  137.09  133.08  134.08     134.08  10025400   
1365  12/14/2022  133.78  135.71  132.76  133.41     133.41   9966100   
1366  12/15/2022  131.15  132.08  129.05  130.10     130.10  12087800   
1367  12/16/2022  129.15  129.90  128.43  129.29     129.29  21126400   

      Daily Return  
0             0.00  
1            -0.93  
2            -0.35  
3             0.14  
4             0.00  
...            ...  
1363          1.55  
1364         -0.10  
1365         -0.50  
1366         -2.48  
1367         -0.62  

[1368 rows x 8 columns]

In [19]:
amazon_df.describe().round(2)

Open     High      Low    Close        Volume  Adj Close  \
count  1250.00  1250.00  1250.00  1250.00  1.250000e+03    1250.00   
mean    120.15   121.59   118.54   120.07  8.670803e+07     120.07   
std      35.66    36.04    35.23    35.59  4.077817e+07      35.59   
min      58.60    59.50    58.53    59.45  1.762600e+07      59.45   
25%      89.21    89.90    88.10    89.13  5.897078e+07      89.13   
50%     108.30   111.70   106.85   108.74  7.515300e+07     108.74   
75%     158.39   160.05   156.28   158.11  1.027046e+08     158.11   
max     187.20   188.65   184.84   186.57  3.113460e+08     186.57   

       Daily Return  
count       1250.00  
mean           0.06  
std            2.25  
min          -14.05  
25%           -1.05  
50%            0.12  
75%            1.17  
max           13.54

We can see the minimum percentage change is a negative 14.05% ie a decrease of 14.05% and the maximum change in the stock prices
is 13.54% in a day ie an increase of 13.54% in a day. Now we do the same for the JPM stock prices

In [20]:
jpm_df.describe().round(2)

Open     High      Low    Close  Adj Close       Volume  \
count  1368.00  1368.00  1368.00  1368.00    1368.00      1368.00   
mean    120.67   121.88   119.46   120.65     112.18  14413609.94   
std      21.82    21.91    21.71    21.81      23.82   6503745.66   
min      81.56    83.75    76.91    79.03      72.58   3220500.00   
25%     104.73   105.70   103.73   104.76      93.90  10228700.00   
50%     114.52   115.40   113.44   114.54     102.74  12855450.00   
75%     134.89   136.45   133.94   135.29     128.46  16357600.00   
max     172.71   172.96   170.54   171.78     166.52  54418800.00   

       Daily Return  
count       1368.00  
mean           0.05  
std            1.96  
min          -14.96  
25%           -0.83  
50%            0.00  
75%            0.96  
max           18.01

Here the decrease in stock prices were similar to that of the amazon stock prices with an decrease of approximately 15% in a day whereas the increase is 18.01% in one of the days.
Also note that the data for JPM stock prices are from July 2017 whereas the data for the amazon stock prices are available from Feb 2018

In [21]:
#Importing the visualization libraries
!pip install seaborn
!pip install plotly
import matplotlib.pyplot as plt

import seaborn as sb
import plotly as px


Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable


In [22]:
#Plotly express plot
fig1 = px.line(title = "Amazon Stock Adjusted Closing Price($) since 2018")
fig1.add_scatter(x = amazon_df['Date'], y = amazon_df['Adj Close'], name = 'Adj Close')

AttributeError: module 'plotly' has no attribute 'line'

In [ ]:
fig2 = px.line(title = "JP Morgan Stock Adjusted Closing Price($) since 2017")
fig2.add_scatter(x = jpm_df['Date'], y = jpm_df['Adj Close'], name = 'JPM Adj Close')

In [ ]:
#A function that performs interactive data visualization using Plotly Express
def plot_fin_data(df, title):
    fig = px.line(title = title)

    #for loop to plot all stock prices and skipping date
    for i in df.columns[1:]:
        fig.add_scatter(x = df['Date'], y = df[i], name = i)
        fig.update_traces(line_width = 5)
        fig.update_layout({'plot_bgcolor' : 'white'})

    fig.show()

In [ ]:
#plotting High, Low, Open, Close
plot_fin_data(amazon_df.drop(['Volume', 'Daily Return', 'Adj Close'], axis = 1), 'Amazon Stock Prices - High, Low, Opening and Closing Price ($)')

In [ ]:
#using iloc to select particular index columns
plot_fin_data(amazon_df.iloc[:,[0,5]], 'Amazon Stocks volume traded since 2018 (in millions)')

In [ ]:
plot_fin_data(amazon_df.iloc[:,[0,7]], 'Amazon Stock Prices - Daily Percentage Returns')

In [ ]:
plot_fin_data(jpm_df.drop(['Adj Close', 'Volume', 'Daily Return'], axis = 1), 'JP Morgan Stock Prices - High, Low, Opening and Closing Prices ($)')

In [ ]:
plot_fin_data(jpm_df.iloc[:,[0,6]], 'JP Morgan Stock Trading Volume (in millions)')

In [ ]:
plot_fin_data(jpm_df.iloc[:,[0,7]], 'JP Morgan Stock - Daily Percentage Returns')

In [ ]:
#function to classify the percentage changes into categories
def perc_change_classifier(percentage):
    if -0.3 <= percentage <= 0.3:
        return "Insignificant Change"
    elif 0.3 < percentage <= 3:
        return "Positive Change"
    elif -3 <= percentage < -0.3:
        return "Negative Change"
    elif 3 < percentage <= 7:
        return "Large Positive Change"
    elif -7 <= percentage < -3:
        return "Large Negative Change"
    elif percentage > 7:
        return "Bull Run"
    elif percentage < -7: 
        return "Bear Sell Off"

In [ ]:
amazon_df['Trend'] = amazon_df['Daily Return'].apply(perc_change_classifier)
jpm_df['Trend'] = jpm_df['Daily Return'].apply(perc_change_classifier)

In [ ]:
amazon_df

In [ ]:
jpm_df

In [ ]:
amazon_trend_summary = amazon_df['Trend'].value_counts()
amazon_trend_summary

In [ ]:
jpm_trend_summary = jpm_df['Trend'].value_counts()
jpm_trend_summary

In [ ]:
#Using matplotlib library to create a pie chart for the trend summaries
plt.figure(figsize = (7,7))
amazon_trend_summary.plot(kind = 'pie', y = 'Trend', ylabel ="")

In [ ]:
plt.figure(figsize = (7,7))
jpm_trend_summary.plot(kind = 'pie', y = 'Trend', ylabel = "")

In [ ]:
#Cufflinks connects pandas and plotly for interactive data visualizations
!pip install cufflinks

import cufflinks as cuff
cuff.go_offline() #for enabling interactive data visualizations

In [ ]:
amazon_df.set_index(['Date'], inplace = True)
jpm_df.set_index(['Date'], inplace = True)

In [ ]:
amazon_df

In [ ]:
jpm_df

In [ ]:
candle_amzn = cuff.QuantFig(amazon_df, title = 'Amazon Inc. AMZN stock - Candlestick chart', name = 'AMZN')
candle_amzn.add_sma(periods = [14,21], column = 'Close', color = ['magenta', 'green'])
candle_amzn.iplot(theme = 'white', up_color = 'green', down_color = 'red')